[Rules] Improve rules matching (#2660)

TD-er opened the pull request:
Detecting time compare and value compare could give false matches in some cases.
This PR also incorporates a number of other changes:
- Adds option to change the remote unit on remote fed plugins via p2p. (#2679)
- Cleanup of use of taskIndex, DeviceIndex (#2676)
- Disable and mark configured plugins not included in the build (e.g. when switching builds with other plugins included) (related to #2676)
- Fix build issues with PlatformIO using Python3
- Perform first read of ADC at init, before starting WiFi (#2646)

Fixes: #2683 - Duplicated section in `platformio.ini`
Fixes: #2681 - sendtohttp malformed request
Fixes: #2660 - json update of Domoticz does not work anymore
Fixes: #2676 - Bug in determine index of Plugin_ptr
Fixes: #2670 - Build error P089
Fixes: #2655 - [int#1] does add a leading space in rules
Fixes: #2650 - AP Mode not working when WIFISSID is set
Fixes: #2686 - Argument parsing does not use a single function
Fixes: #2657 - [Rules] Parsing of [task#var] syntax changed since 20190928

TD-er committed:
[Rules] Improve rules matching (#2660)

Detecting time compare and value compare could give false matches in some cases.

TD-er committed:
[PIO] Fix build issue using Python 3

PlatformIO 4.0.x does have issues with non-ASCII characters in files. Even if they are part of comments.

TD-er committed:
[Rules] Make handling compare condition more deterministic.

TD-er committed:
[PIO] Fix build issue due to duplicate env in platformio.ini

TD-er committed:
[Cleanup] Make use of consistent names for DeviceIndex (#2676)

See #2676
The use of DeviceIndex was not consistent, which may lead to strange issues.
DeviceIndex is the translation between the vector of included plugins and IDs of plugins available.

TD-er committed:
[Cleanup] Proper handling of tasks not supported in current build

When using settings containing configuration for a certain plugin which is not included in the actual build, then the plugin must be displayed as "not supported" and not possible to edit.
Also show some indication about the consequences of trying to edit such a task. It will then present a dialog equal to adding a new device to a task.

TD-er committed:
[ESPEasy p2p] Allow changing remote unit on receiving plugin (#2679)

See also #2679

TD-er committed:
[Cleanup] Use of taskIndex_t & reduce memory usage in structs

Some structs like controller queue elements had their members not sorted to size. This may result in more memory usage than needed.

TD-er committed:
[Cleanup] Use taskIndex_t everywhere where a task index is used.

TD-er committed:
[sendToHttp] Trim spaces from path (#2681, #2660) 

The changed rules parsing may have added some spaces in replaced variables.
Adding a trim() call to the parseString does remove any existing or added spaces in the parameter.

TD-er committed:
[WDT reboot] Fix SW watchdog reboot when saving rules

The ESP could sometimes experience a crash when saving the rules.
This was caused mainly on fragmented file systems and a bit longer rules files.
Now the same saveToFile function is used as being used in all other settings file access.
Also the rules were copied in memory, which is useless since they are already present in memory (LWIP  code) and passed by const reference.

TD-er committed:
[ADC] Perform first read of ADC at init, before starting WiFi (#2646)

See #2646

TD-er committed:
[ArduinoIDE] Fix include errors for ArduinoIDE

As noted by https://github.com/letscontrolit/ESPEasy/issues/2681#issuecomment-546306710  @clumsy-stefan

clumsy-stefan commented:
If I understand correctly, Rules still chekc only for the first occurence of a specifig task name and then if the value in the rule is in this task. Therefore it's not possible to have multiple tasks with the same name to be used in rules.

Would'nt it be more logical (and quite easy to solve) if the rules would continue to search for a task with the same name if in the first (previous) occurence no value name that matches is found.

As an example: I would want all possible systemvalues in one common "taskname" (for some controllers this means "the same device") eg. sysinfo but as I'm limited to 4 per task I need multiple tasks for all values. Same for dummy-values. However I can't use them anymore in Rules hen, as only the first occurence is checked..

Just a question/suggestion

TD-er commented:
A lot of code does make the assumption a task name is unique.
I also added an error message a while ago when you try to save a task with a name that is already present.

The variable names in a task may also be re-used for other tasks.
The true constraint must be that the `[taskname#varname]` is unique.
Currently we have it a bit more strict to also have the restriction that the task name is unique, since it will become very confusing otherwise.

For example:
- task1: `foo`, variable `barone`
- task1: `foo`, variable `bartwo`

It just does not make sense to have multiple tasks with the same name.
If you need it for administrating in controllers, then we may need to add a system variable like `%pluginid%` or `%pluginname%` and not make the handling of addressing a specific value more ambiguous.

clumsy-stefan commented:
>The true constraint must be that the [taskname#varname] is unique.

Yep, this is also what I would expect, so it would cover all cases...

>It just does not make sense to have multiple tasks with the same name.

hmm.. I think this is POV thing. for me it's totally reasonable... eg. if I need more than 4 dummy vars, why should I identify them via task-name and not via variable name (from hierarchical POV it just makes more sense)...

But again IMHO and I can live with how it's now... as long as it's only a "warning" and no restirction. it's then just not possible to use a variable in a task with duplicat name (in rules)...

TD-er commented:
Like I said, if you need it for administrative purposes in a controller to set the MQTT topic for example, we may need to have the plugin ID resolved in the controller.

I'm not 100% sure if we can also match an event on only the task name.
If that cannot be done, then it may be an option to have the restriction to have `[taskname#varname]` unique.
But changing that will not be an easy or quick fix.
For example it would mean there is no longer an 1-to-1 relation anymore between task and task-name and that does have quite a lot of consequences.

clumsy-stefan commented:
hmm.. but in the rules when you check for `[task#value]` you only check in the first occurence of `task` if a coresponding `value`in there is not found it just returnes (false). instead of stop checking rules then you could continue cycling through the tasks until `TASK_MAX` instead of stop checking (so basically jsut finish the loop if the value is not found).

TD-er commented:
Hmm if that's true, then that is a bug.
An event should match all what is defined in the event.

- `on taskname#* do`  should trigger on any event starting with `taskname#`
- `on taskname#varname do`  should trigger only on events matching `taskname#varname`

Still, the asterisk could give the false impression on which plugin was updated and you will then use a variable from another plugin which was not yet updated.

clumsy-stefan commented:
I always check for `task#value` but it only matches if the corresponding `value` is in first occurence of `task`.

Eg: if I have 2 tasks witch are called `dummy` and have `value[1-4]` in the first and `vlaue[5-8]` in the second I can never match for `dummy#vlaue5`.

TD-er commented:
Ah OK, I get it.
That has to do with the searching for task names indeed.
Not sure about the implications when I would extend the search.
For sure the caching of the search for task names has to be re-written, since you would then require a 1-to-many relation.

Maybe you can add an issue for it so we can continue the discussion there instead of at the comments section of a pull request which will be merged one day :)

TD-er committed:
[WiFi AP]  No automatic attempt to reconnect with connected client (#2650)

See #2650.
As long as a client is connected to the node in AP mode, no automatic reconnect should be attempted since it will make it very hard to get connected or remain connected to the AP mode.
The WiFi channel will change during scan which will disconnect the user.

TD-er committed:
[WiFi] Only reset wifi connect attempt on stable connection (#2650)

See #2650
The counter wifi_connect_attempt is being used to determine whether the fallback SSID should be tried.
It is also used to determine whether a "quick reconnect" is possible using the BSSID and channel of the last successful connection.
If the connection is unstable, (last connection was active for over 5 minutes) then the connect attempt counter can be reset.

TD-er committed:
[MQTT] Use parseString functions to disect publish command (#2657)

See #2657

TD-er committed:
[Rules] Fix grammar in parsing parameters (#2686)

See: #2686

TD-er committed:
[Commands] Mark commands not using standard parse functions (#2686)

TD-er committed:
[PIO] Cleanup old environments and add debug definitions

To be used for the static analysis tools of PlatformIO.
See [announcement PIO Home 3.0 beta](https://community.platformio.org/t/platformio-home-3-0-beta-and-powerful-project-inspection/10187?utm_source=platformio&utm_medium=piohome)
